multimodal/src/autogluon/multimodal/models/sam.py

```diff
@@ -2,20 +2,278 @@
+    # Concatenate output tokens
+    output_tokens = torch.cat([self.iou_token.weight, self.mask_tokens.weight], dim=0)
+    output_tokens = output_tokens.repeat(batch_size, point_batch_size, 1, 1)
+
+    if sparse_prompt_embeddings.sum().item() != 0:
+        tokens = torch.cat((output_tokens, sparse_prompt_embeddings), dim=2)
+    else:
+        tokens = output_tokens
+    point_embeddings = tokens.to(self.iou_token.weight.dtype)
+
+    # Expand per-image data in batch direction to be per-point
+    image_embeddings = image_embeddings + dense_prompt_embeddings
+    image_embeddings = image_embeddings.repeat(point_batch_size, 1, 1, 1)
+    image_positional_embeddings = image_positional_embeddings.repeat(point_batch_size, 1, 1, 1)
+
+    # Run the transformer, image_positional_embedding are consumed
+    point_embedding, image_embeddings, attentions = self.transformer(
+        point_embeddings=point_embeddings,
+        image_embeddings=image_embeddings,
+        image_positional_embeddings=image_positional_embeddings,
+        attention_similarity=attention_similarity,
+        target_embedding=target_embedding,
+        output_attentions=output_attentions,
+    )
+    iou_token_out = point_embedding[:, :, 0, :]
+    mask_tokens_out = point_embedding[:, :, 1 : (1 + self.num_mask_tokens), :]
+
+    # Upscale mask embeddings and predict masks using the mask tokens
+    image_embeddings = image_embeddings.transpose(2, 3).reshape(
+        batch_size * point_batch_size, num_channels, height, width
+    )
+
+    upscaled_embedding = self.upscale_conv1(image_embeddings)
+    upscaled_embedding = self.activation(self.upscale_layer_norm(upscaled_embedding))
+    upscaled_embedding = self.activation(self.upscale_conv2(upscaled_embedding))
+
+    # new added
+    hyper_in = self.output_hypernetworks_mlps[0](mask_tokens_out)  # bs, num_channels, num_token, 32
+
+    _, num_channels, height, width = upscaled_embedding.shape
+    upscaled_embedding = upscaled_embedding.reshape(batch_size, point_batch_size, num_channels, height * width)
+    masks = (hyper_in @ upscaled_embedding).reshape(
+        batch_size, point_batch_size, -1, height, width
+    )  # bs, 1, num_tokens, h, w
+    class_predictions = self.output_classifier_mlps(mask_tokens_out).reshape(
+        batch_size, point_batch_size, -1, self.num_classes + 1
+    )  # new added
+
+    # Generate mask quality predictions
+    iou_pred = self.iou_prediction_head(iou_token_out)
+
+    # # Select the correct mask or masks for output
+    # if multimask_output:
+    #     mask_slice = slice(1, None)
+    # else:
+    #     mask_slice = slice(0, 1)
+    # masks = masks[:, :, mask_slice, :, :]
+    # iou_pred = iou_pred[:, :, mask_slice]
+
+    outputs = (masks, iou_pred)
+
+    if output_attentions:
+        outputs = outputs + (attentions,)
+    else:
+        outputs = outputs + (None,)
+
+    return outputs + (class_predictions,)
+
+
+def multi_class_sam_model_forward(
```

Comment:
Inside the function, better to mark which are original code and which are new added.

Reply:
Add the original link of the copied code.

Reply:
I think a better way would be keeping the commented old code in contrast to the new ones.